pkg/storagev2/inmemchunkstore/inmemchunkstore.go

```diff
@@ -13,49 +14,119 @@ import (
 )
 
 type ChunkStore struct {
-	chunks sync.Map
+	chunks map[string]chunkData
+	lock   sync.Mutex
+}
+
+type chunkData struct {
+	chunk  swarm.Chunk
+	stamps []swarm.Stamp
 }
 
 func New() *ChunkStore {
-	return &ChunkStore{}
+	return &ChunkStore{
+		chunks: make(map[string]chunkData),
+	}
+}
+
+func (c *ChunkStore) Get(ctx context.Context, addr swarm.Address) (swarm.Chunk, error) {
+	return c.GetWithStamp(ctx, addr, nil)
 }
 
-func (c *ChunkStore) Get(_ context.Context, addr swarm.Address) (swarm.Chunk, error) {
-	val, found := c.chunks.Load(addr.ByteString())
-	if !found {
+func (c *ChunkStore) GetWithStamp(ctx context.Context, addr swarm.Address, batchID []byte) (swarm.Chunk, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	data, ok := c.chunks[addr.ByteString()]
+	if !ok {
 		return nil, storage.ErrNotFound
 	}
-	return val.(swarm.Chunk), nil
+
+	// when batchID is not specified, chunk with first stamp is returned
+	if batchID == nil {
+		return makeChunk(data.chunk, data.stamps[0]), nil
```

Comment:
Before dereferencing the stamps array we should check if there is any stamp present.

I think it is better to have 2 errors in storage package:
ErrNoStampsForChunk => if no stamp is present
ErrStampNotFound => returned by GetWithStamp if stamp is not found

As both these cases are different. So more information is better.